### ISSUE TITLE
Recent introduction of __del__ exposes a bug in Python 3.4 when requests library is used

### ISSUE THREAD

**Issue by vvolkman** [django-haystack/pysolr#193]
This class of bug was first [Reported in the requests library:](url)
(https://github.com/kennethreitz/requests/issues/2303)

What I do here is insert a print("entering __del__") right in the __del__ method destructor of pysolr library.  Every time a program or thread exits, a ReferenceError results.

Remove the __del__ destructor and no issues.

```
$ python3
Python 3.4.3 (default, Apr  1 2015, 18:10:40)
[GCC 4.8.2 20140120 (Red Hat 4.8.2-16)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pysolr
>>> s = pysolr.Solr("del sucks")
>>> exit()
**entering __del__**
Exception ignored in: <bound method Solr.__del__ of <pysolr.Solr object at 0x7fac7c9408d0>>
Traceback (most recent call last):
  File "/opt/pqapps/alerting/pysolr.py", line 329, in __del__
  File "/usr/local/lib/python3.4/site-packages/requests/sessions.py", line 648, in close
  File "/usr/lib64/python3.4/_collections_abc.py", line 512, in __iter__
  File "/usr/lib64/python3.4/collections/__init__.py", line 91, in __iter__
ReferenceError: weakly-referenced object no longer exists
```

Here's what happens when I comment out the whole __del__ dtor

```
$ python3
Python 3.4.3 (default, Apr  1 2015, 18:10:40)
[GCC 4.8.2 20140120 (Red Hat 4.8.2-16)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pysolr
>>> s = pysolr.Solr("del sucks")
>>> exit()
```

In other words, business as usual

**Comment by acdha** [comment]
That's unfortunate – the `__del__` method to tear down sessions was needed to reduce other logspam . We could work around that in the unittest suite by explicitly closing the sessions first but that won't help actual apps.

**Comment by acdha** [comment]
2eb7371fe2fdfc4321e3a88407a673b3ba7c2436 should take care of the other problem but we have a problem with the SolrCloud support added in #138 due to the way Kazoo uses threads to retry failed ZooKeeper connections forever and will thus prevent a python app from exiting cleaning before the server becomes functional again.

df2b3f3741c4b53fb467c32cacb14bd477c91663 seems to be the simplest way to address this by setting a max retry limit so things like tests will exit eventually.
